test/FullRange.t.sol

```diff
@@ -0,0 +1,453 @@
+
+        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 14546694553059925434 - 1000);
+        assertEq(owed, true);
+    }
+
+    function testTwoSwaps() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+
+        IPoolManager.SwapParams memory params =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 1 ether, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+        PoolSwapTest.TestSettings memory settings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        snapStart("swap first");
+        swapRouter.swap(key, params, settings);
+        snapEnd();
+
+        (bool owed,) = fullRange.poolInfo(id);
+        assertEq(owed, true);
+
+        snapStart("swap second");
+        swapRouter.swap(key, params, settings);
+        snapEnd();
+
+        (owed,) = fullRange.poolInfo(id);
+        assertEq(owed, true);
+    }
+
+    function testSwapAddLiquidityTwoPools() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+        manager.initialize(key2, SQRT_RATIO_1_1);
+
+        fullRange.addLiquidity(address(token0), address(token1), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+        fullRange.addLiquidity(address(token1), address(token2), 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+
+        IPoolManager.SwapParams memory params =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 10000000, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+
+        PoolSwapTest.TestSettings memory testSettings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        swapRouter.swap(key, params, testSettings);
+        swapRouter.swap(key2, params, testSettings);
+
+        (bool owed,) = fullRange.poolInfo(id);
+        assertEq(owed, true);
+
+        (owed,) = fullRange.poolInfo(id2);
+        assertEq(owed, true);
+    }
+
+    function testInitialRemoveLiquiditySucceeds() public {
+        manager.initialize(key, SQRT_RATIO_1_1);
+
+        uint256 prevBalance0 = MockERC20(token0).balanceOf(address(this));
+        uint256 prevBalance1 = MockERC20(token1).balanceOf(address(this));
+
+        address token0Addr = address(token0);
+        address token1Addr = address(token1);
+
+        fullRange.addLiquidity(token0Addr, token1Addr, 3000, 10 ether, 10 ether, address(this), MAX_DEADLINE);
+
+        (, address liquidityToken) = fullRange.poolInfo(id);
+
+        assertEq(UniswapV4ERC20(liquidityToken).balanceOf(address(this)), 10 ether - 1000);
+
+        assertEq(MockERC20(token0).balanceOf(address(this)), prevBalance0 - 10 ether);
+        assertEq(MockERC20(token1).balanceOf(address(this)), prevBalance1 - 10 ether);
+
+        UniswapV4ERC20(liquidityToken).approve(address(fullRange), type(uint256).max);
+
+        snapStart("remove liquidity");
```

Comment:
fyi since remove liquidity is technically in the same transaction these are all dirty storage slots so unfortunately is likely downplaying actual gas...Im not sure how to get around this other than adding liquidity in some pool that's used specifically for remove liquidty gas tests 🥱. Not super important to address for now... did something similar here... https://github.com/Uniswap/v4-core/blob/main/test/foundry-tests/TickBitmap.t.sol#L19